File: src/main/java/com/fengsheng/skill/QiangLing.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.phase.OnChooseReceiveCard;
import com.fengsheng.phase.OnSendCard;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * 张一挺技能【强令】：你传出情报后，或你决定接收情报后，可以宣言至多两个卡牌名称。本回合中，所有角色均不能使用被宣言的卡牌。
 */
public class QiangLing implements TriggeredSkill {
    @Override
    public SkillId getSkillId() {
        return SkillId.QIANG_LING;
    }

    @Override
    public ResolveResult execute(Game g) {
        if (g.getFsm() instanceof OnSendCard fsm) {
            Player r = fsm.whoseTurn();
            if (r.findSkill(getSkillId()) == null)
                return null;
            if (r.getSkillUseCount(getSkillId()) >= 1)
                return null;
            r.addSkillUseCount(getSkillId());
            return new ResolveResult(new executeQiangLing(fsm, r), true);
        } else if (g.getFsm() instanceof OnChooseReceiveCard fsm) {
            Player r = fsm.inFrontOfWhom();
            if (r.findSkill(getSkillId()) == null)
                return null;
            if (r.getSkillUseCount(getSkillId()) >= 2)
                return null;
            r.addSkillUseCount(getSkillId(), 2);
            return new ResolveResult(new executeQiangLing(fsm, r), true);
        }
        return null;
    }

    private record executeQiangLing(Fsm fsm, Player r) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeQiangLing.class);

        @Override
        public ResolveResult resolve() {
            for (Player player : r.getGame().getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    var builder = Role.skill_wait_for_qiang_ling_toc.newBuilder();
                    builder.setPlayerId(p.getAlternativeLocation(r.location()));
                    builder.setWaitingSecond(20);
                    if (p == r) {
                        final int seq2 = p.getSeq();
                        builder.setSeq(seq2);
                        GameExecutor.post(p.getGame(), () -> {
                            if (p.checkSeq(seq2))
                                p.getGame().tryContinueResolveProtocol(p, Role.skill_qiang_ling_tos.newBuilder().setEnable(false).setSeq(seq2).build());
                        }, p.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS);
                    }
                    p.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer) {
                GameExecutor.post(r.getGame(), () -> {
                    List<Common.card_type> result = new ArrayList<>();
                    if (r.getGame().getQiangLingTypes().isEmpty()) {
                        var cardTypes = new Common.card_type[]{Common.card_type.Jie_Huo, Common.card_type.Diao_Bao, Common.card_type.Wu_Dao};
                        int idx = ThreadLocalRandom.current().nextInt(cardTypes.length);
                        for (int i = 0; i < cardTypes.length; i++) {
                            if (i != idx)
                                result.add(cardTypes[i]);
                        }
                    } else {
                        var cardTypes = EnumSet.of(Common.card_type.Jie_Huo, Common.card_type.Diao_Bao, Common.card_type.Wu_Dao);
                        r.getGame().getQiangLingTypes().forEach(cardTypes::remove);
                        result.add(Common.card_type.Po_Yi);
                        for (var t : cardTypes) {
                            result.add(t);
                            break;
                        }
                    }
                    r.getGame().tryContinueResolveProtocol(r, Role.skill_qiang_ling_tos.newBuilder().setEnable(true).addAllTypes(result).build());
                }, 2, TimeUnit.SECONDS);
            }
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != r) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_qiang_ling_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (!pb.getEnable()) {
                r.incrSeq();
                return new ResolveResult(fsm, true);
            }
            if (pb.getTypesCount() == 0) {
                log.error("enable为true时types不能为0");
                return null;
            }
            for (var t : pb.getTypesList()) {
                if (t == Common.card_type.UNRECOGNIZED || t == null) {
                    log.error("未知的卡牌类型" + t);
                    return null;
                }
            }
            r.incrSeq();
            log.info(r + "发动了[强令]，禁止了" + Arrays.toString(pb.getTypesList().toArray(new Common.card_type[0])));
            r.getGame().getQiangLingTypes().addAll(pb.getTypesList());
            for (Player p : r.getGame().getPlayers()) {
                if (p instanceof HumanPlayer player1)
                    player1.send(Role.skill_qiang_ling_toc.newBuilder()
                            .setPlayerId(player1.getAlternativeLocation(r.location()))
                            .addAllTypes(pb.getTypesList()).build());
            }
            return new ResolveResult(fsm, true);
        }
    }

    public static void resetQiangLing(Game game) {
        game.getQiangLingTypes().clear();
    }
}

File: src/main/kotlin/skill/QiangLing.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.protos.Common.card_type.*
import com.fengsheng.protos.Role.skill_qiang_ling_tos
import com.fengsheng.protos.skillQiangLingToc
import com.fengsheng.protos.skillQiangLingTos
import com.fengsheng.protos.skillWaitForQiangLingToc
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 张一挺技能【强令】：你传出情报后，或你决定接收情报后，可以宣言至多两个卡牌名称。本回合中，所有角色均不能使用被宣言的卡牌。
 */
class QiangLing : TriggeredSkill {
    override val skillId = SkillId.QIANG_LING

    override val isInitialSkill = true

    override fun execute(g: Game, askWhom: Player): ResolveResult? {
        val event1 = g.findEvent<SendCardEvent>(this) { event ->
            askWhom === event.sender
        }
        if (event1 != null)
            return ResolveResult(ExecuteQiangLing(g.fsm!!, event1, askWhom), true)
        val event2 = g.findEvent<ChooseReceiveCardEvent>(this) { event ->
            askWhom === event.inFrontOfWhom
        }
        if (event2 != null) {
            return ResolveResult(ExecuteQiangLing(g.fsm!!, event2, askWhom), true)
        }
        return null
    }

    private data class ExecuteQiangLing(val fsm: Fsm, val event: Event, val r: Player) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            r.game!!.players.send { player ->
                skillWaitForQiangLingToc {
                    playerId = player.getAlternativeLocation(r.location)
                    waitingSecond = Config.WaitSecond
                    if (player === r) {
                        val seq2 = player.seq
                        seq = seq2
                        player.timeout = GameExecutor.post(player.game!!, {
                            if (player.checkSeq(seq2))
                                player.game!!.tryContinueResolveProtocol(player, skillQiangLingTos { seq = seq2 })
                        }, player.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) {
                GameExecutor.post(r.game!!, {
                    if (event is ChooseReceiveCardEvent) {
                        val v = r.calculateMessageCardValue(event.whoseTurn, r, event.messageCard, sender = event.sender)
                        if (v < 0) { // 如果情报不值得接收，就不发动强令
                            r.game!!.tryContinueResolveProtocol(r, skillQiangLingTos { })
                            return@post
                        }
                    }
                    val result = listOf(Jie_Huo, Diao_Bao, Wu_Dao)
                        .filterNot { r.cannotPlayCard(it) }.run {
                            when (size) {
                                0 -> listOf(Po_Yi, Cheng_Qing).filterNot { r.cannotPlayCard(it) }
                                1 -> plus(if (event is SendCardEvent && !r.cannotPlayCard(Po_Yi)) Po_Yi else Cheng_Qing)
                                2 -> this
                                else -> sortedBy { type -> r.cards.any { it.type == type } }.take(2)
                            }
                        }
                    r.game!!.tryContinueResolveProtocol(r, skillQiangLingTos {
                        enable = result.isNotEmpty()
                        types.addAll(result)
                    })
                }, 3, TimeUnit.SECONDS)
            }
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== r) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_qiang_ling_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (!message.enable) {
                r.incrSeq()
                return ResolveResult(fsm, true)
            }
            if (message.typesCount == 0) {
                logger.error("enable为true时types不能为空")
                player.sendErrorMessage("[强令]的卡牌类型不能为空")
                return null
            }
            val typesList = message.typesList.toList()
            for (t in typesList) {
                if (t == UNRECOGNIZED || t == null) {
                    logger.error("未知的卡牌类型$t")
                    player.sendErrorMessage("未知的卡牌类型$t")
                    return null
                }
            }
            r.incrSeq()
            logger.info("${r}发动了[强令]，禁止了${typesList.joinToString()}")
            r.game!!.players.forEach { it!!.skills += CannotPlayCard(cardType = typesList) }
            r.game!!.players.send {
                skillQiangLingToc {
                    playerId = it.getAlternativeLocation(r.location)
                    types.addAll(typesList)
                }
            }
            return ResolveResult(fsm, true)
        }
    }
}
